### PR TITLE
WT-9418 Introduce the ref track build option

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
@agorrod After this change, I need to make changes to the server scons script to enable the flag.

**Review by agorrod — COMMENTED** [review]
Looks good to me, it deserves testing via MongoDB before merging, even though it feels safe to me.

I think this should work without change in MongoDB when it's first merged, so I don't see a need to co-ordinate. OTOH: I agree that it would be worth understanding how this would work in the server before merging, to make sure it's worth doing at all.

Holding off on my LGTM until the ancillary work is done, but code wise this change looks fine to me.

**Comment by quchenhao** [comment]
@agorrod mongodb patch build looks ok.

**Review by agorrod — APPROVED** [review]
LGTM - could you find one other reviewer to please?

File: cmake/configs/base.cmake
Comment (ajmorton) [inline cmake/configs/base.cmake:288]: Is `HAVE_REF_TRACK_DISABLED` dead code? I don't see it used anywhere

File: cmake/README.md
Comment (ajmorton) [inline cmake/README.md:101]: It's not clear what ref tracking is from this text. Maybe "Enable WiredTiger to track recent state transitions for WT_REF structures (always enabled in diagnostic build)"?
The same can be done in base.cmake if we make this change.

File: src/include/btmem.h
Comment (ajmorton) [inline src/include/btmem.h:1033]: I think this comment should still say DIAGNOSTIC mode. I don't see anything in this diff that indicates `HAVE_REF_TRACK` will zero out freed WT_REF memory.

File: cmake/configs/base.cmake
Comment (quchenhao) [reply cmake/configs/base.cmake:288]: It's a cmake internal argument. Not used in c code.

```
# config_bool(config_name description DEFAULT <default-value> [DEPENDS <deps>] [DEPENDS_ERROR <config-val> <error-string>])
# Defines a boolean (ON/OFF) configuration option . The configuration option is stored in the cmake cache
# and can be exported to the wiredtiger config header.
#   config_name - name of the configuration option.
#   description - docstring to describe the configuration option (viewable in the cmake-gui).
#   DEFAULT <default-value> -  default value of the configuration bool (ON/OFF). Used when not manually set
#       by a cmake script or in the cmake-gui or when dependencies aren't met.
#   DEPENDS <deps> - list of dependencies (semicolon seperated) required for the configuration bool
#       to be set to the desired value. If any of the dependencies aren't met the configuration value
#       will be set to its default value.
#   DEPENDS_ERROR <config-val> <error-string> - specifically throw a fatal error when the configuration option is set to
#       <config-val> despite failing on its dependencies. This is mainly used for commandline-like options where you want
#       to signal a specific error to the caller when dependencies aren't met e.g. toolchain is missing library (as opposed to
#       silently defaulting).
function(config_bool config_name description)
    cmake_parse_arguments(
        PARSE_ARGV
        2
        "CONFIG_BOOL"
        ""
        "DEFAULT;DEPENDS"
        "DEPENDS_ERROR"
    )

    if(NOT "${CONFIG_BOOL_UNPARSED_ARGUMENTS}" STREQUAL "")
        message(FATAL_ERROR "Unknown arguments to config_bool: ${CONFIG_BOOL_UNPARSED_ARGUMENTS}")
    endif()
    # We require a default value (not optional).
    if("${CONFIG_BOOL_DEFAULT}" STREQUAL "")
        message(FATAL_ERROR "No default value passed")
    endif()

    set(depends_err_value)
    set(depends_err_message "")
    # If DEPENDS_ERROR is specifically set, parse the value we want to throw an error on if the dependency fails.
    if(CONFIG_BOOL_DEPENDS_ERROR)
        list(LENGTH CONFIG_BOOL_DEPENDS_ERROR depends_error_length)
        if(NOT depends_error_length EQUAL 2)
            message(FATAL_ERROR "Invalid usage of DEPENDS_ERROR: requires <Error Value> <Error Message>")
        else()
            list(GET CONFIG_BOOL_DEPENDS_ERROR 0 err_val)
            if(err_val)
                set(depends_err_value "1")
            else()
                set(depends_err_value "0")
            endif()
            list(GET CONFIG_BOOL_DEPENDS_ERROR 1 depends_err_message)
        endif()
    endif()

    # Check that the configs dependencies are enabled before setting it to a visible enabled state.
    eval_dependency("${CONFIG_BOOL_DEPENDS}" enabled)
    if(enabled)
        # Set an internal cache variable "${config_name}_DISABLED" to capture its enabled/disabled state.
        # We want to ensure we capture a transition from a disabled to enabled state when dependencies are met.
        if(${config_name}_DISABLED)
            unset(${config_name}_DISABLED CACHE)
            set(${config_name} ${CONFIG_BOOL_DEFAULT} CACHE STRING "${description}" FORCE)
        else()
            set(${config_name} ${CONFIG_BOOL_DEFAULT} CACHE STRING "${description}")
        endif()
    else()
        set(config_value "0")
        if (${${config_name}})
            set(config_value "1")
        endif()
        # If the user tries to set the config option to a given value when its dependencies
        # are not met, throw an error (when DEPENDS_ERROR is explicitly set).
        if(CONFIG_BOOL_DEPENDS_ERROR)
            if(${depends_err_value} EQUAL ${config_value})
                message(FATAL_ERROR "Unable to set ${config_name}: ${depends_err_message}")
            endif()
        endif()
        # Config doesn't meet dependency requirements, set its default state and flag it as disabled.
        set(${config_name} OFF CACHE STRING "${description}" FORCE)
        set(${config_name}_DISABLED ON CACHE INTERNAL "" FORCE)
    endif()
endfunction()
```

File: src/include/btmem.h
Comment (quchenhao) [reply src/include/btmem.h:1033]: I don't follow this comment. Now the ref track structure is only enabled with HAVE_REF_TRACK not DIAGNOSTIC flag. It can be enabled in production build as well.

**Review by ajmorton — APPROVED** [review]
LGTM!

**Comment by quchenhao** [comment]
evergreen merge

**Pull request merged**
